File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        for(auto &c: s){
            if(c== '(' || c=='{' || c=='[')
                stack.push(c);
            else if(stack.empty())
                return false;
            else if(c=='}' && stack.top() != '{')
                return false;
            else if(c==')' && stack.top() != '(')
                return false;
            else if(c==']' && stack.top() != '[')
                return false;
            else
                stack.pop();
        }
        return stack.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        compliment = {')':'(', '}':'{',']':'[' }
        for c in s:
            if c in compliment:
                if len(stack):
                    if compliment[c] == stack[-1]:
                        stack.pop()
                    else:
                        return False
                else:
                    return False
            else:
                stack.append(c)
        return len(stack) == 0